File: DAM1/Programacion/eclipse-workspace/pooProtectora/Cliente.java
package pooProtectora;

import java.util.Arrays;

public class Cliente {
	String nombre;
	String apellidos;
	int edad;
	int telefono;
	String interes;

	Gato[] gatos = null;
	Perro[] perros = null;
	Tortuga tortuga = null;

	public Cliente(String nombre, String apellidos, int edad, int telefono, String interes) {
		this.nombre = nombre;
		this.apellidos = apellidos;
		this.edad = edad;
		this.telefono = telefono;
		this.interes = interes;
	}

	// cuando un cliente adopta un perro primero comprobamos si no ha sobrepasado ya los dos máximo
	// si no ha sido así lo marcamos como adoptado y referenciamos el animal en los arrays que tiene cada cliente
	public void adopta(Perro perro) {
		if (perros!=null && perros.length == 2)
			System.out.println("Este cliente ya ha adoptado dos perros");
		else {
			perro.adoptado();
			if (perros == null) {
				perros = new Perro[1];
				perros[0] = perro;
			} else{
				perros = Arrays.copyOf(perros, perros.length + 1);
				perros[perros.length - 1] = perro;
			}
		}
	}

	// El resto de animales es igual
	public void adopta(Gato gato) {
		if (gatos!=null && gatos.length == 3)
			System.out.println("Este cliente ya ha adoptado tres gatos");
		else {
			gato.adoptado();
			if (gatos == null) {
				gatos = new Gato[1];
				gatos[0] = gato;
			} else{
				gatos = Arrays.copyOf(gatos, gatos.length + 1);
				gatos[gatos.length - 1] = gato;
			}
		}
	}

	// En este caso si usamos this para diferenciar el atributo del objeto del argumento de la función
	// ya que se llaman igual
	public void adopta(Tortuga tortuga) {
		if (this.tortuga == null) {
			tortuga.adoptado();
			this.tortuga = tortuga;
		}
		else
			System.out.println("\nEste cliente ya ha adoptado una tortuga");
	}

	// Para listar los animales adoptados por un cliente recorremos los arrays (la tortuga es un dato simple)
	// y llamamos a la función que muestra sus datos en consola que ya usamos en el listado de la protectora
	public void listaAdoptados() {
		System.out.println("\nCliente: " + nombre + " " + apellidos + " (" + telefono + ")");
		if(perros!=null) {
			System.out.println("Perros adoptados:");
			for(Perro perro:this.perros)
				perro.datos();
		}
		if(gatos!=null) {
			System.out.println("Gatos adoptados:");
			for(Gato gato:this.gatos)
				gato.datos();
		}
		if(tortuga!=null) {
			System.out.println("Tortuga adoptada:");
			tortuga.datos();
		}
	}
}

File: DAM1/Phyton/PycharmProjects1/PrimehProyecto/clases/cliente.py
class Cliente:

    def __init__(self, nombre, apellido, edad, telefono):
        self.nombre = nombre
        self.apellido = apellido
        self.edad = edad
        self.telefono = telefono

        self.animales = []

    def adoptar(self, animal):

        perros = 0
        gatos = 0
        tortugas = 0

        for a in self.animales:
            if a.tipo == "perro":
                perros += 1
            elif a.tipo == "gato":
                gatos += 1
            elif a.tipo == "tortuga":
                tortugas += 1


        if len(self.animales) >= 4:
            print("No puedes adoptar más de 4 animales")
            return

        if animal.tipo == "perro" and perros >= 2:
            print("No puedes adoptar más perros")
            return

        if animal.tipo == "gato" and gatos >= 3:
            print("No puedes adoptar más gatos")
            return

        if animal.tipo == "tortuga" and tortugas >= 1:
            print("No puedes adoptar más tortugas")
            return

        self.animales.append(animal)
        print("Animal adoptado correctamente")
    
    
    def mostrar_animales(self):
        print("Animales de", self.nombre)
        if len(self.animales) == 0:
            print("No ha adoptado ningún animal")
        else:
            for a in self.animales:
              a.mostrar()